File: packages/sdk-ts/src/wrappers/_event.ts
// Shared event-factory for LLM wrappers. Fills all the constant fields so
// each wrapper only specifies what it knows: provider, model, token counts,
// latency, status, optional tool_name, optional token_count_source.

import { randomUUID } from 'node:crypto';
import {
  CostSource,
  EventStatus,
  Framework,
  InstrumentationTier,
  Provider,
  TokenCountSource,
  type TelemetryEvent,
} from '@pylva/shared';
import { currentContext } from '../core/context.js';

export interface BuildLlmEventInput {
  provider: string | null;
  model: string | null;
  tokensIn: number;
  tokensOut: number;
  latencyMs: number;
  status: EventStatus;
  tokenCountSource?: TokenCountSource | undefined;
  stepNameFallback?: string | undefined;
  toolName?: string | null | undefined;
}

// Returns the shape the telemetry exporter expects — `Omit<TelemetryEvent, 'schema_version' | 'sdk_version'>`.
export function buildLlmEvent(
  input: BuildLlmEventInput,
): Omit<TelemetryEvent, 'schema_version' | 'sdk_version'> {
  const ctx = currentContext();
  return {
    run_id: ctx?.run_id ?? randomUUID(),
    parent_run_id: ctx?.parent_run_id ?? null,
    trace_id: ctx?.trace_id ?? randomUUID(),
    span_id: randomUUID(),
    parent_span_id: ctx?.span_id ?? null,
    customer_id: ctx?.customer_id ?? 'anonymous',
    step_name: ctx?.step_name ?? input.stepNameFallback ?? null,
    model: input.model,
    provider: input.provider ?? Provider.OTHER,
    tokens_in: input.tokensIn,
    tokens_out: input.tokensOut,
    latency_ms: input.latencyMs,
    tool_name: input.toolName ?? null,
    status: input.status,
    framework: ctx?.framework ?? Framework.NONE,
    instrumentation_tier: InstrumentationTier.SDK_WRAPPER,
    cost_source: CostSource.AUTO,
    metric: null,
    metric_value: null,
    stream_aborted: input.status === EventStatus.ABORTED,
    abort_savings_usd: 0,
    timestamp: new Date().toISOString(),
    ...(input.tokenCountSource !== undefined
      ? { metadata: { token_count_source: input.tokenCountSource } }
      : {}),
  };
}

File: packages/sdk-py/pylva/wrappers/_event.py
"""Shared event-factory for LLM wrappers — mirrors TS wrappers/_event.ts.

Fills the constant fields so each wrapper only specifies what it knows.
"""

from __future__ import annotations

import uuid
from datetime import datetime, timezone
from typing import Any

from ..core.context import current_context
from ..core.identifiers import clean_provider_model_identifier


def _utc_now_iso() -> str:
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


def build_llm_event(
    *,
    provider: Any | None,
    model: Any | None,
    tokens_in: int,
    tokens_out: int,
    latency_ms: int,
    status: str,                 # success | failure | aborted
    token_count_source: str | None = None,  # exact | estimated
    tool_name: str | None = None,
    step_name_fallback: str | None = None,
) -> dict[str, Any]:
    """Build a TelemetryEvent-shaped dict ready for enqueue()."""
    ctx = current_context()
    safe_provider = (
        clean_provider_model_identifier(provider) if provider is not None else "other"
    )
    event: dict[str, Any] = {
        "run_id": ctx.run_id if ctx else str(uuid.uuid4()),
        "parent_run_id": ctx.parent_run_id if ctx else None,
        "trace_id": ctx.trace_id if ctx else str(uuid.uuid4()),
        "span_id": str(uuid.uuid4()),
        "parent_span_id": ctx.span_id if ctx else None,
        "customer_id": ctx.customer_id if ctx else "anonymous",
        "step_name": (ctx.step_name if ctx else None) or step_name_fallback,
        "model": clean_provider_model_identifier(model),
        "provider": safe_provider,
        "tokens_in": tokens_in,
        "tokens_out": tokens_out,
        "latency_ms": latency_ms,
        "tool_name": tool_name,
        "status": status,
        "framework": ctx.framework if ctx else "none",
        "instrumentation_tier": "sdk_wrapper",
        "cost_source": "auto",
        "metric": None,
        "metric_value": None,
        "stream_aborted": status == "aborted",
        "abort_savings_usd": 0,
        "timestamp": _utc_now_iso(),
    }
    if token_count_source is not None:
        event["metadata"] = {"token_count_source": token_count_source}
    return event
